Azure/sonic-buildimage
# Added HTTP server support on chassis system

Signed-off-by: Anand Mehra <anamehra@cisco.com>

#### Why I did it


This commit introduces HTTP server for Supervisor Card. HTTP will be used by Line Cards to download Sonic image
file from Supervisor card. This may also be used to download configuration files or any other data file from Supervisor.

#### How I did it

HTTP server is enabled by a systemd service, http-server.service and runs on Supervisor only when
/etc/sonic/http_config.yaml file is present.

On Supervisor Card, the platform shall provide HTTP configuration via devices/chassis.yaml file.
Following HTTP configuration may be provided by chassis.yaml:

http-config:
    http_server_dir: <http server home dir path>
    http_server_ip: <ip>
    http_server_port: <port>
    start_http_server: 'yes'

On Line Card, chassis.yaml shall only contain:
http-config:
    http_server_ip: 127.1.1.1

http_server_dir:         Optional; HTTP home directory, default: /var/www
http_server_ip:        Required; IP to map http-server hostname used by HTTP Server.
                                  This should be the internal IP address used by the internal chassis network.
http_server_port:    Optional; Port to bind, default: 8000.
start_http_server:   Optional; If 'yes' or '1', start HTTP server on the card.

chassis.yaml file is parsed by a new service, config-chassis.service. This service parses the http-config
from chassis.yaml to set up pre-requisites for http-server.service. config-chassis.service is made generic
to be used for other chassis-based configurations in the future.

config-chassis.service does the following two operations for HTTP server based on http-config block:

If start_http_server is 'yes' or '1', it creates /etc/sonic/http_config.yaml file with the HTTP configuration
provided.
Create the http_server_ip to http-server hostname mapping in /etc/hosts. Line Cards may also provide
http-config with this entry only to create hostname mapping during boot.

#### How to verify it
create devices/<platform>/chassis.yaml file with following details:
http-config:
    http_server_dir: <http server home dir path>
    http_server_ip: <ip>
    http_server_port: <port>
    start_http_server: 'yes'

During boot config-chassis.servcei should parse this file and create /etc/hosts entry for 'http-server' hostname
and /etc/sonic/http_config.yaml

http-server.service should start and create http server.

Place a file at the http_server_dir path and use curl to download:
curl http://http-server:<http_server_port>/<file> -o file

validate the file.

[UT.txt](https://github.com/Azure/sonic-buildimage/files/7188900/UT.txt)

#### Which release branch to backport (provide reason below if selected)




#### Description for the changelog



#### A picture of a cute animal (not mandatory but encouraged)
